File: frontend/afc-frontend/src/api/tracker.ts
import { apiRequest } from "./apiClient";
import type { OrderDetailPayload } from "./ordersTable";

export type OutgoingOrderType = "Installation" | "Will Call" | "Shipment" | "Delivery";

export type Department =
  | "SALES"
  | "LOGISTICS"
  | "DELIVERY_DEPT"
  | "SERVICE"
  | "ACCOUNTING";

export interface OrderTrackerPayload {
  id: number;
  order_id: number;
  current_department: Department;
  step_index: number;
  is_backordered: boolean;
  updated_at: string;
}

export interface OrderHistoryPayload {
  id: number;
  order_id: number;
  from_department: Department | null;
  to_department: Department;
  action_taken: string;
  performed_by: string;
  completed_at: string;
  comments?: string | null;
}

export interface OrderTrackerStagePayload {
  id: number;
  order_id: number;
  stage_index: number;
  is_completed: boolean;
  completed_by: string | null;
  completed_at: string | null;
}

/** Joined view: an Order with its current tracking status and full history. */
export interface OrderWithTracking {
  order: OrderDetailPayload;
  tracker: OrderTrackerPayload | null;
  history: OrderHistoryPayload[];
  stages: OrderTrackerStagePayload[];
}

/** Single row returned by the GET /packing-slips endpoint. */
export interface PackingSlipResult {
  id: number;
  order_number: string;
  external_order_number?: string | null;
  order_type?: string | null;
  status: string;
  description?: string | null;
  customer_name?: string | null;
  supplier_name?: string | null;
  created_at: string;
  completed_at?: string | null;
  eta?: string | null;
  is_paid: boolean;
  is_invoiced: boolean;
  tracker: OrderTrackerPayload | null;
  history: OrderHistoryPayload[];
  stages: OrderTrackerStagePayload[];
}

export interface PackingSlipsResponse {
  page: number;
  limit: number;
  total: number;
  status_counts: {
    "Not Started": number;
    "In Progress": number;
    Completed: number;
    Backordered: number;
  };
  results: PackingSlipResult[];
}

export interface TrackerFilters{
  page?: number;
  limit?: number;
  search?: string;
  tracker_status?: string;
  tracker_department?: string | string[];
  order_type?: string | string[];
  party?: string | string[];
  // Created date filters
  start_date?: string;
  end_date?: string;
  before_date?: string;
  after_date?: string;
  // Last updated date filters
  updated_start_date?: string;
  updated_end_date?: string;
  updated_before_date?: string;
  updated_after_date?: string;
}

function appendListParam(query: URLSearchParams, key: string, values?: string | string[]) {
  const list = Array.isArray(values) ? values : values ? [values] : [];
  for (const value of list) {
    const trimmed = value.trim();
    if (trimmed) query.append(key, trimmed);
  }
}

export function fetchOrderTracking(orderId: number | string) {
  return apiRequest(`/orders/${orderId}/tracker`) as Promise<OrderWithTracking>;
}

export function initOrderTracker(
  orderId: number | string,
  payload: { current_department: Department; step_index?: number }
) {
  return apiRequest(`/orders/${orderId}/tracker`, {
    method: "POST",
    body: JSON.stringify(payload),
  }) as Promise<OrderTrackerPayload>;
}

export function updateOrderTracker(
  orderId: number | string,
  payload: { current_department?: Department; step_index?: number; is_backordered?: boolean }
) {
  return apiRequest(`/orders/${orderId}/tracker`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  }) as Promise<OrderTrackerPayload>;
}

export function toggleTrackerStage(
  orderId: number | string,
  stageIndex: number,
  payload: { is_completed: boolean; completed_by?: string; department?: Department }
) {
  return apiRequest(`/orders/${orderId}/tracker/stages/${stageIndex}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  }) as Promise<OrderTrackerStagePayload>;
}

export function addOrderHistory(
  orderId: number | string,
  payload: {
    from_department?: Department | null;
    to_department: Department;
    action_taken: string;
    performed_by: string;
    comments?: string | null;
  }
) {
  return apiRequest(`/orders/${orderId}/history`, {
    method: "POST",
    body: JSON.stringify(payload),
  }) as Promise<OrderHistoryPayload>;
}

export function patchOrderPaidInvoiced(
  orderId: number | string,
  payload: { is_paid?: boolean; is_invoiced?: boolean }
) {
  return apiRequest(`/orders/${orderId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function fetchPackingSlips(filters?: TrackerFilters): Promise<PackingSlipsResponse> {
  const query = new URLSearchParams();
  if (filters?.page) query.set("page", String(filters?.page));
  if (filters?.limit) query.set("limit", String(filters?.limit));
  if (filters?.search) query.set("search", filters?.search);
  if (filters?.tracker_status) query.set("tracker_status", filters?.tracker_status);
  appendListParam(query, "tracker_department", filters?.tracker_department);
  appendListParam(query, "order_type", filters?.order_type);
  appendListParam(query, "party", filters?.party);
  // Created date filters
  if (filters?.start_date) query.set("start_date", filters?.start_date);
  if (filters?.end_date) query.set("end_date", filters?.end_date);
  if (filters?.before_date) query.set("before_date", filters?.before_date);
  if (filters?.after_date) query.set("after_date", filters?.after_date);
  // Last updated date filters
  if (filters?.updated_start_date) query.set("updated_start_date", filters?.updated_start_date);
  if (filters?.updated_end_date) query.set("updated_end_date", filters?.updated_end_date);
  if (filters?.updated_before_date) query.set("updated_before_date", filters?.updated_before_date);
  if (filters?.updated_after_date) query.set("updated_after_date", filters?.updated_after_date);
  return apiRequest(`/packing-slips?${query.toString()}`) as Promise<PackingSlipsResponse>;
}

File: frontend/afc-frontend/src/api/tracker.py
from flask import Blueprint, g, jsonify, request
from flask_jwt_extended import jwt_required, get_jwt
from sqlalchemy import select, func, or_, case, and_
from sqlalchemy.exc import IntegrityError, DatabaseError
from database.models import Order, OrderTracker, OrderHistory, OrderTrackerStage, Department, OutgoingOrderType, Customer, Supplier, OrderType, OUTGOING_TYPES
from datetime import datetime, timezone
from typing import Tuple, Any

from app.api.error_handling import (
    handle_database_error,
    safe_commit,
    ResourceNotFoundError,
)

# All order types that participate in the packing-slip tracker
# (outgoing types + incoming / purchase orders)
TRACKER_TYPES = OUTGOING_TYPES | {OrderType.INCOMING.value}

# Maps each Department value to the permission required to update it.
DEPARTMENT_PERMISSION_MAP = {
    Department.SALES.value: "tracker:update_sales",
    Department.SERVICE.value: "tracker:update_service",
    Department.LOGISTICS.value: "tracker:update_logistics",
    Department.DELIVERY_DEPT.value: "tracker:update_delivery",
}

# SQLAlchemy CASE expression: total stages expected for each order type
# Must mirror the frontend step-path definitions (INSTALLATION_STEPS, WILL_CALL_STEPS, PURCHASE_ORDER_STEPS)
_total_steps_expr = case(
    (Order.type == OrderType.INSTALLATION.value, 6),
    (Order.type.in_([OrderType.WILL_CALL.value, OrderType.DELIVERY.value, OrderType.SHIPMENT.value]), 4),
    else_=3,  # incoming / purchase order (and legacy "outgoing")
)

def _check_department_permission(user_permissions, department):
    """Return an error response if the user lacks permission for the given department, or None if allowed."""
    required_perm = DEPARTMENT_PERMISSION_MAP.get(department)
    if required_perm and required_perm not in user_permissions:
        return jsonify({"error": "Forbidden: You do not have permission to update this department."}), 403
    return None


tracker_bp = Blueprint("tracker", __name__)


@tracker_bp.route("/orders/<int:order_id>/tracker", methods=["GET"])
@jwt_required()
def get_order_tracker(order_id: int) -> Tuple[Any, int]:
    """Return the current tracking state and full history for an order."""
    db = g.db

    order = db.get(Order, order_id)
    if not order:
        raise ResourceNotFoundError("Order", order_id)

    tracker = db.execute(
        select(OrderTracker).where(OrderTracker.order_id == order_id)
    ).scalar_one_or_none()

    history_rows = db.execute(
        select(OrderHistory)
        .where(OrderHistory.order_id == order_id)
        .order_by(OrderHistory.completed_at.asc())
    ).scalars().all()

    stage_rows = db.execute(
        select(OrderTrackerStage)
        .where(OrderTrackerStage.order_id == order_id)
        .order_by(OrderTrackerStage.stage_index.asc())
    ).scalars().all()

    return jsonify({
        "order": order.to_dict(),
        "tracker": tracker.to_dict() if tracker else None,
        "history": [h.to_dict() for h in history_rows],
        "stages": [s.to_dict() for s in stage_rows],
    }), 200


@tracker_bp.route("/orders/<int:order_id>/tracker", methods=["POST"])
@jwt_required()
def create_order_tracker(order_id: int) -> Tuple[Any, int]:
    """Initialize tracking for an order (sets current_department and step_index)."""
    db = g.db
    data = request.get_json() or {}

    order = db.get(Order, order_id)
    if not order:
        raise ResourceNotFoundError("Order", order_id)

    existing = db.execute(
        select(OrderTracker).where(OrderTracker.order_id == order_id)
    ).scalar_one_or_none()
    if existing:
        return jsonify({"error": "Tracker already exists for this order."}), 409

    current_department = data.get("current_department")
    if not current_department or current_department not in [d.value for d in Department]:
        return jsonify({
            "error": "current_department must be one of: " + ", ".join(d.value for d in Department)
        }), 400

    tracker = OrderTracker(
        order_id=order_id,
        current_department=current_department,
        step_index=data.get("step_index", 0),
        warehouse_id=g.active_warehouse_id,
        updated_at=datetime.now(timezone.utc),
    )
    db.add(tracker)

    error = safe_commit(db)
    if error:
        return handle_database_error(error)

    return jsonify(tracker.to_dict()), 201


@tracker_bp.route("/orders/<int:order_id>/tracker", methods=["PATCH"])
@jwt_required()
def update_order_tracker(order_id: int) -> Tuple[Any, int]:
    """Advance the tracker to a new department/step, set backordered flag, or mark invoiced/paid."""
    db = g.db
    data = request.get_json() or {}

    user_permissions = get_jwt().get("permissions", [])

    # Permission check for department updates
    target_department = data.get("current_department")
    if target_department:
        denied = _check_department_permission(user_permissions, target_department)
        if denied:
            return denied

    # Permission check for marking as invoiced
    if "is_invoiced" in data and "orders:mark_invoiced" not in user_permissions:
        return jsonify({"error": "Forbidden: You do not have permission to mark orders as invoiced."}), 403

    # Permission check for marking as paid
    if "is_paid" in data and "orders:mark_paid" not in user_permissions:
        return jsonify({"error": "Forbidden: You do not have permission to mark orders as paid."}), 403

    order = db.get(Order, order_id)
    if not order:
        raise ResourceNotFoundError("Order", order_id)

    tracker = db.execute(
        select(OrderTracker).where(OrderTracker.order_id == order_id)
    ).scalar_one_or_none()
    if not tracker:
        return jsonify({"error": "Tracker not found for this order."}), 404

    current_department = data.get("current_department")
    if current_department is not None:
        if current_department not in [d.value for d in Department]:
            return jsonify({
                "error": "current_department must be one of: " + ", ".join(d.value for d in Department)
            }), 400
        tracker.current_department = current_department

    if "step_index" in data:
        tracker.step_index = data["step_index"]

    if "is_backordered" in data:
        is_backordered = data["is_backordered"]
        if not isinstance(is_backordered, bool):
            return jsonify({"error": "is_backordered must be a boolean."}), 400
        tracker.is_backordered = is_backordered

    if "is_invoiced" in data:
        is_invoiced = data["is_invoiced"]
        if not isinstance(is_invoiced, bool):
            return jsonify({"error": "is_invoiced must be a boolean."}), 400
        order.is_invoiced = is_invoiced

    if "is_paid" in data:
        is_paid = data["is_paid"]
        if not isinstance(is_paid, bool):
            return jsonify({"error": "is_paid must be a boolean."}), 400
        order.is_paid = is_paid

    tracker.updated_at = datetime.now(timezone.utc)

    error = safe_commit(db)
    if error:
        return handle_database_error(error)

    return jsonify(tracker.to_dict()), 200


@tracker_bp.route("/orders/<int:order_id>/history", methods=["POST"])
@jwt_required()
def add_order_history(order_id: int) -> Tuple[Any, int]:
    """Append a history entry (department transition + action) for an order."""
    db = g.db
    data = request.get_json() or {}

    user_permissions = get_jwt().get("permissions", [])
    to_dept = data.get("to_department")
    from_dept = data.get("from_department")
    if to_dept:
        denied = _check_department_permission(user_permissions, to_dept)
        if denied:
            return denied
    if from_dept:
        denied = _check_department_permission(user_permissions, from_dept)
        if denied:
            return denied

    order = db.get(Order, order_id)
    if not order:
        raise ResourceNotFoundError("Order", order_id)

    department_values = [d.value for d in Department]

    to_department = data.get("to_department")
    if not to_department or to_department not in department_values:
        return jsonify({
            "error": "to_department must be one of: " + ", ".join(department_values)
        }), 400

    from_department = data.get("from_department")
    if from_department and from_department not in department_values:
        return jsonify({
            "error": "from_department must be one of: " + ", ".join(department_values)
        }), 400

    action_taken = data.get("action_taken", "").strip()
    if not action_taken:
        return jsonify({"error": "action_taken is required."}), 400

    performed_by = data.get("performed_by", "").strip()
    if not performed_by:
        return jsonify({"error": "performed_by is required."}), 400

    entry = OrderHistory(
        order_id=order_id,
        from_department=from_department,
        to_department=to_department,
        action_taken=action_taken,
        performed_by=performed_by,
        completed_at=datetime.now(timezone.utc),
        comments=data.get("comments"),
    )
    db.add(entry)

    error = safe_commit(db)
    if error:
        return handle_database_error(error)

    return jsonify(entry.to_dict()), 201


@tracker_bp.route("/orders/<int:order_id>/tracker/stages/<int:stage_index>", methods=["PATCH"])
@jwt_required()
def toggle_tracker_stage(order_id: int, stage_index: int) -> Tuple[Any, int]:
    """Toggle the completion state of a specific tracker stage for an order."""
    db = g.db
    data = request.get_json() or {}

    user_permissions = get_jwt().get("permissions", [])
    target_department = data.get("department")
    if target_department:
        denied = _check_department_permission(user_permissions, target_department)
        if denied:
            return denied

    order = db.get(Order, order_id)
    if not order:
        raise ResourceNotFoundError("Order", order_id)

    is_completed = data.get("is_completed")
    if is_completed is None or not isinstance(is_completed, bool):
        return jsonify({"error": "is_completed (boolean) is required."}), 400

    completed_by = data.get("completed_by", "").strip() or None

    stage = db.execute(
        select(OrderTrackerStage).where(
            OrderTrackerStage.order_id == order_id,
            OrderTrackerStage.stage_index == stage_index,
        )
    ).scalar_one_or_none()

    if stage is None:
        stage = OrderTrackerStage(
            order_id=order_id,
            stage_index=stage_index,
            is_completed=is_completed,
            completed_by=completed_by if is_completed else None,
            completed_at=datetime.now(timezone.utc) if is_completed else None,
        )
        db.add(stage)
    else:
        stage.is_completed = is_completed
        stage.completed_by = completed_by if is_completed else None
        stage.completed_at = datetime.now(timezone.utc) if is_completed else None

    # Clear the backordered flag whenever a stage is toggled
    tracker = db.execute(
        select(OrderTracker).where(OrderTracker.order_id == order_id)
    ).scalar_one_or_none()
    if tracker and tracker.is_backordered:
        tracker.is_backordered = False
        tracker.updated_at = datetime.now(timezone.utc)

    error = safe_commit(db)
    if error:
        return handle_database_error(error)

    return jsonify(stage.to_dict()), 200


@tracker_bp.route("/packing-slips", methods=["GET"])
@jwt_required()
def get_packing_slips() -> Tuple[Any, int]:
    """Return all tracker-eligible orders (outgoing + purchase/incoming) with their tracker and history info."""
    db = g.db

    page = request.args.get("page", default=1, type=int)
    limit = request.args.get("limit", default=25, type=int)
    search = request.args.get("search", "").strip()
    tracker_status = request.args.get("tracker_status", "").strip()  # "Not Started"|"In Progress"|"Completed"|"Backordered"
    order_type = request.args.get("order_type", "").strip()
    offset = (page - 1) * limit

    # Correlated sub-query: count of completed stages for each order row
    _completed_stages_subq = (
        select(func.count(OrderTrackerStage.id))
        .where(
            OrderTrackerStage.order_id == Order.id,
            OrderTrackerStage.is_completed == True,  # noqa: E712
        )
        .correlate(Order)
        .scalar_subquery()
    )

    # Status conditions (mirror the frontend toPackingSlipRow logic)
    _backordered_cond = OrderTracker.is_backordered == True  # noqa: E712
    _not_started_cond = and_(OrderTracker.id.is_(None), _completed_stages_subq == 0)
    _completed_cond = and_(
        _completed_stages_subq >= _total_steps_expr,
        or_(OrderTracker.id.is_(None), OrderTracker.is_backordered == False),  # noqa: E712
    )
    _in_progress_cond = and_(
        or_(OrderTracker.id.is_not(None), _completed_stages_subq > 0),
        _completed_stages_subq < _total_steps_expr,
        or_(OrderTracker.id.is_(None), OrderTracker.is_backordered == False),  # noqa: E712
    )

    # Base query: all tracker-eligible orders joined with customer, supplier and tracker
    base_query = (
        select(Order)
        .outerjoin(Customer, Order.customer_id == Customer.id)
        .outerjoin(Supplier, Order.supplier_id == Supplier.id)
        .outerjoin(OrderTracker, OrderTracker.order_id == Order.id)
        .where(Order.type.in_(TRACKER_TYPES))
        .where(Order.warehouse_id == g.active_warehouse_id)
    )

    if search:
        base_query = base_query.where(
            or_(
                Order.order_number.ilike(f"%{search}%"),
                Order.external_order_number.ilike(f"%{search}%"),
                Customer.name.ilike(f"%{search}%"),
                Supplier.name.ilike(f"%{search}%"),
            )
        )

    if order_type:
        base_query = base_query.where(Order.type == order_type)

    # Tracker status filter using stage-completion counts
    if tracker_status == "Not Started":
        base_query = base_query.where(_not_started_cond)
    elif tracker_status == "In Progress":
        base_query = base_query.where(_in_progress_cond)
    elif tracker_status == "Completed":
        base_query = base_query.where(_completed_cond)
    elif tracker_status == "Backordered":
        base_query = base_query.where(_backordered_cond)

    # Efficient total count using the filtered query as a subquery
    subq = base_query.subquery()
    total = db.execute(
        select(func.count()).select_from(subq)
    ).scalar()

    orders = db.execute(
        base_query.order_by(Order.created_at.desc()).limit(limit).offset(offset)
    ).scalars().all()

    # Per-status counts for the tab badges (based on search, ignoring tracker_status)
    _search_filter = (
        or_(
            Order.order_number.ilike(f"%{search}%"),
            Order.external_order_number.ilike(f"%{search}%"),
            Customer.name.ilike(f"%{search}%"),
            Supplier.name.ilike(f"%{search}%"),
        )
        if search
        else None
    )
    counts_query = (
        select(
            func.count(case((_backordered_cond, 1))).label("backordered"),
            func.count(case((_not_started_cond, 1))).label("not_started"),
            func.count(case((_in_progress_cond, 1))).label("in_progress"),
            func.count(case((_completed_cond, 1))).label("completed"),
        )
        .select_from(Order)
        .outerjoin(Customer, Order.customer_id == Customer.id)
        .outerjoin(Supplier, Order.supplier_id == Supplier.id)
        .outerjoin(OrderTracker, OrderTracker.order_id == Order.id)
        .where(Order.type.in_(TRACKER_TYPES))
        .where(Order.warehouse_id == g.active_warehouse_id)
    )
    if _search_filter is not None:
        counts_query = counts_query.where(_search_filter)
    if order_type:
        counts_query = counts_query.where(Order.type == order_type)
    counts_row = db.execute(counts_query).one()
    status_counts = {
        "Not Started": counts_row.not_started or 0,
        "In Progress": counts_row.in_progress or 0,
        "Completed": counts_row.completed or 0,
        "Backordered": counts_row.backordered or 0,
    }

    results = []
    for order in orders:
        tracker = order.tracker
        history = sorted(order.history, key=lambda h: h.completed_at)
        stages = sorted(order.stages, key=lambda s: s.stage_index)
        results.append({
            "id": order.id,
            "order_number": order.order_number,
            "external_order_number": order.external_order_number,
            "order_type": order.type,
            "status": order.status,
            "description": order.description,
            "customer_name": order.customer.name if order.customer else None,
            "supplier_name": order.supplier.name if order.supplier else None,
            "created_at": order.created_at.isoformat() if order.created_at else None,
            "completed_at": order.completed_at.isoformat() if order.completed_at else None,
            "eta": order.eta.strftime("%Y-%m-%d") if order.eta else None,
            "is_paid": order.is_paid,
            "is_invoiced": order.is_invoiced,
            "tracker": tracker.to_dict() if tracker else None,
            "history": [h.to_dict() for h in history],
            "stages": [s.to_dict() for s in stages],
        })

    return jsonify({
        "page": page,
        "limit": limit,
        "total": total,
        "status_counts": status_counts,
        "results": results,
    }), 200
